zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/util/XmlUtil.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.gateway.rest.util;
+
+import io.camunda.search.entities.FlowNodeInstanceEntity;
+import io.camunda.search.entities.UserTaskEntity;
+import io.camunda.service.ProcessDefinitionServices;
+import io.camunda.zeebe.protocol.record.value.BpmnElementType;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+import org.xml.sax.SAXException;
+
+@Component
+public class XmlUtil {
+
+  public static final String XPATH_NODE_NAME = "//process[@id='%s']/%s[@id='%s']/@name";
+  private static final Logger LOG = LoggerFactory.getLogger(XmlUtil.class);
+
+  private final ProcessDefinitionServices processDefinitionServices;
+
+  public XmlUtil(final ProcessDefinitionServices processDefinitionServices) {
+    this.processDefinitionServices = processDefinitionServices;
+  }
+
+  public String getFlowNodeName(final FlowNodeInstanceEntity flowNode) {
+    final var processDefinition =
+        processDefinitionServices.getByKey(flowNode.processDefinitionKey());
+    final var type = BpmnElementType.valueOf(flowNode.type().name());
+    if (type.getElementTypeName().isEmpty()) {
+      return flowNode.flowNodeId();
+    }
+    return extractFlowNodeName(
+        processDefinition.bpmnXml(),
+        processDefinition.processDefinitionId(),
+        type.getElementTypeName().get(),
+        flowNode.flowNodeId());
+  }
+
+  public String getUserTaskName(final UserTaskEntity userTask) {
+    final var processDefinition =
+        processDefinitionServices.getByKey(userTask.processDefinitionKey());
+    return extractFlowNodeName(
+        processDefinition.bpmnXml(),
+        processDefinition.processDefinitionId(),
+        "userTask",
+        userTask.elementId());
+  }
+
+  public String extractFlowNodeName(
+      final String xml,
+      final String processDefinitionId,
+      final String flowNodeType,
+      final String flowNodeId) {
+    try {
+      final var documentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
```

Comment:
💭 Would be nicer to extract these factories. However, these classes don't seem to be thread-safe according to spec: [DocumentBuilderFactory](https://docs.oracle.com/cd/E17802_01/webservices/webservices/docs/1.5/api/javax/xml/parsers/DocumentBuilderFactory.html#:~:text=provided%20the%20instance%20of%20the%20factory%20isn%27t%20being%20used%20in%20more%20than%20one%20thread%20at%20a%20time), [DocumentBuilder](https://docs.oracle.com/cd/E17802_01/webservices/webservices/docs/1.5/api/javax/xml/parsers/DocumentBuilder.html#:~:text=is%20NOT%20guaranteed%20to%20behave%20as%20per%20the%20specification%20if%20it%20is%20used%20concurrently%20by%20two%20or%20more%20threads).


Define them as `ThreadLocal` to make it faster?